### ISSUE TITLE
Magic compatibility with Psychtoolbox

### ISSUE THREAD

**Issue by Stefanatuurlijk** [nigelrogasch/MAGIC#18]
Hi all,

I am having some issues regarding the compatibility with the Psychtoolbox. I am currently trying to set up an experiment where a visual stimulus is presented to the subject, whereafter a particular time a TMS train of pulses needs to be given. I first wrote two separate scripts for the Magic Toolbox and the Psychtoolbox, and those work wonderfully, but when I combine the two the TMS gets disconnected from my PC. I know this because the TMS gets disarmed when I open a screen using the Psychtoolbox, but Matlab thinks it is still connected and armed. Is there something I can do to prevent this disconnection while using the Psychtoolbox and make it so that I can trigger the TMS during the visual stimuli? 

Kind Regards,
Stefan.

**Comment by umair-hassan** [comment]
Hi Stefan, 

in order to best reproduce the results can you share your matlab version, psychtoolbox version? may be attach a simple script that I can use to reproduce your problem?

I think its the sequence of steps, if you initialize  magic class after opening the screen, and do it in a different workspace structure, then it could work

**Comment by Stefanatuurlijk** [comment]
Hi Umair-hassan,

Thanks for the quick response,

I am using Matlab 2023a with the latest version of Magic toolbox and version 3.0.19 of the Psychtoolbox. Right now I use my Windows laptop to run the code but in the future I will need to use a windows computer with multiple monitors.

Below is the code that I use;

p.s. I have some experience with programming in Matlab, but I still can still learn a lot more!

%% initiate TMS, connect to the Rapid2 stimulator
unlockCode = 'xxxx-xxxxxxxx-xx';
myMS = rapid('COM7','rapid',unlockCode);
myMS.connect();
Amplitude =35;
myMS.setAmplitudeA(Amplitude);
myMS.rTMSMode(1);
trainparameters.frequency =[]; 
trainparameters.nPulses =[5]; 
trainparameters.duration =[1]; 
myMS.setTrain(trainparameters);
myMS.setChargeDelay(2);
myMS.arm();

%% Initiate pcyshtoolbox
PsychDefaultSetup(2);
screens = Screen('Screens');
screenNumber = max(screens);
white = WhiteIndex(screenNumber);
black = BlackIndex(screenNumber);
startXpix = 120;
startYpix = 50;
dimX = 400;
dimY = 250;

%% Start the experiment.
[window, windowRect] = PsychImaging('OpenWindow', screenNumber, white...
    ,[startXpix startYpix startXpix + dimX startYpix + dimY]);
ifi = Screen('GetFlipInterval', window);
[screenXpixels, screenYpixels] = Screen('WindowSize', window);
[xCenter, yCenter] = RectCenter(windowRect);
Screen('BlendFunction', window, 'GL_SRC_ALPHA', 'GL_ONE_MINUS_SRC_ALPHA');
presSecs = 1;
waitframes = round(presSecs / ifi);
randomImage = 1:3;
Screen('TextSize', window, 80);
Screen('TextFont', window, 'Courier');
DrawFormattedText(window, 'press any key to start', 'center', 'center', black);
KbStrokeWait;
Screen('Flip', window);
WaitSecs(2);

for i = 1:length(randomImage)
ImageLocation = ['C:\Users\image', sprintf('%d',randomImage(i)), '.png'];
theImage = imread(ImageLocation);
[s1, s2, s3] = size(theImage);
imageTexture = Screen('MakeTexture', window, theImage);
vbl = Screen('Flip', window);
Screen('DrawTexture', window, imageTexture, [], [], 0);
vbl  = Screen('Flip', window);

WaitSecs(0.3);
myMS.fire();

WaitSecs(0.7);
Screen('FillRect', window, white);
vbl  = Screen('Flip', window);

WaitSecs(3);
end

myMS.disarm()
myMS.disconnect()

**Comment by umair-hassan** [comment]
Hi, can you try this code and send the outcomes?

%% initiate TMS, connect to the Rapid2 stimulator


%% Initiate pcyshtoolbox
PsychDefaultSetup(2);
screens = Screen('Screens');
screenNumber = max(screens);
white = WhiteIndex(screenNumber);
black = BlackIndex(screenNumber);
startXpix = 120;
startYpix = 50;
dimX = 400;
dimY = 250;

%% Start the experiment.
[window, windowRect] = PsychImaging('OpenWindow', screenNumber, white...
,[startXpix startYpix startXpix + dimX startYpix + dimY]);

unlockCode = 'xxxx-xxxxxxxx-xx';
myMS = rapid('COM7','rapid',unlockCode);
myMS.connect();
Amplitude =35;
myMS.setAmplitudeA(Amplitude);
myMS.rTMSMode(1);
trainparameters.frequency =[];
trainparameters.nPulses =[5];
trainparameters.duration =[1];
myMS.setTrain(trainparameters);
myMS.setChargeDelay(2);
myMS.arm();


ifi = Screen('GetFlipInterval', window);
[screenXpixels, screenYpixels] = Screen('WindowSize', window);
[xCenter, yCenter] = RectCenter(windowRect);
Screen('BlendFunction', window, 'GL_SRC_ALPHA', 'GL_ONE_MINUS_SRC_ALPHA');
presSecs = 1;
waitframes = round(presSecs / ifi);
randomImage = 1:3;
Screen('TextSize', window, 80);
Screen('TextFont', window, 'Courier');
DrawFormattedText(window, 'press any key to start', 'center', 'center', black);
KbStrokeWait;
Screen('Flip', window);
WaitSecs(2);

for i = 1:length(randomImage)
ImageLocation = ['C:\Users\image', sprintf('%d',randomImage(i)), '.png'];
theImage = imread(ImageLocation);
[s1, s2, s3] = size(theImage);
imageTexture = Screen('MakeTexture', window, theImage);
vbl = Screen('Flip', window);
Screen('DrawTexture', window, imageTexture, [], [], 0);
vbl = Screen('Flip', window);

WaitSecs(0.3);
myMS.fire();

WaitSecs(0.7);
Screen('FillRect', window, white);
vbl = Screen('Flip', window);

WaitSecs(3);
end

myMS.disarm()
myMS.disconnect()

**Comment by umair-hassan** [comment]
i would also try this in different matlab versions for example

- [ ] 2017b

- [ ] 2019a

- [ ] 2021b

**Comment by Stefanatuurlijk** [comment]
I tried the code in 2017b, 2021b and 2023a and the same happens with all Matlab versions, it works until a new screen is opened via the Psychtoolbox when I delete the opening of the screen before the loop, the loop starts and the TMS triggers, but when the screen is flipped, the TMS disconnects again.

**Comment by umair-hassan** [comment]
Okay, there is something in the Psychtoolbox that clears your workspace and disconnects all the devices. Would you like to give it a try with a screenshare online with me?

**Comment by umair-hassan** [comment]
@Stefanatuurlijk the pause is not may be as accurate - caution!

**Comment by Stefanatuurlijk** [comment]
I will keep that in mind, If I find another solution, I will let you know

**Comment by kleinerm** [comment]
Weird that a proper wait on a thread on Windows would cause a serial port to disconnect. I see this toolbox uses Matlabs deprecated `serial()` function for serial i/o, probably a limitation of that function. Might be worth switching to Psychtoolbox `IOPort()` function which is optimized for neuroscience use cases and also has been found to have better timing / lower latency, e.g., for trigger emission.

**Comment by umair-hassan** [comment]
i agree, we need to make a new release that uses non-deprecated serial io func. Thanks for pointing to IOport() @kleinerm

**Comment by TheHandLab** [comment]
jumping in here: I just tested MAGIC `bistim` `fire()` latency on Windows 10, 64bit, Matlab R2022a
i did 100 calls from matlab, 5s apart, recorded at 4kHz on a DAQ box
comparing the NI card's `outputsinglescan` to MAGIC's `fire()` command, the minimum latency between these two calls increasing the voltage on the DAQ was 8.5ms, max was 11.5ms, mean was exactly 10ms. so, this is around 10+-1.5ms for any single call.

is that the expected latency for the COM port? @umair-hassan thanks!

**Comment by umair-hassan** [comment]
@TheHandLab The observed latency of ~10±1.5ms is very good and reassuring for your setup, it is important to note that this cannot be generalized across all systems. Latency depends on various factors, including the specific hardware, Matlab configuration, and system load, PC/mac memory, other Matlab threads running, the calls planned in the loop iterations etc. In some setups, latency can be even faster than this (with even reduced jitter), while in others, it may slow down to 50ms or even 100ms ( often with increased jitter). Documenting these measurements for each setup is important like you did.
